### PR TITLE
Sort order of updates for user properties

### LINKED ISSUE #32852: Prevent deadlocks on concurrent user updates
### Description

The current functionality when setting a user attribute will first remove the user attribute using a named query. 

When doing this, and two operations execute the deleted for the same user in for the attributes in a different order, this can lead to deadlocks. 

See for the current code: 

https://github.com/keycloak/keycloak/blob/37f42430e6b99ac261d8936dbd1f41b9de9bfb3d/model/jpa/src/main/java/org/keycloak/models/jpa/UserAdapter.java#L206-L233

### Discussion

_No response_

### Motivation

While the result of a concurrent update is usually between "last write wins" and a call wouldn't know if they won, we can avoid a deadlock by sorting the updates to ensure a stable ordering. 

### Details

Reproducer: 

Disable auto-commit and aetup the data:

```
insert into user_attribute (name, value, user_id, id) values ('1', 'v', 'c7332040-acfd-41a4-90cd-a61c408b040b', '1');

insert into user_attribute (name, value, user_id, id) values ('2', 'v', 'c7332040-acfd-41a4-90cd-a61c408b040b', '2');
```

Run in session one: 

```
delete from user_attribute where user_id = 'c7332040-acfd-41a4-90cd-a61c408b040b' and name = '1';
```

Run in session two:

```
delete from user_attribute where user_id = 'c7332040-acfd-41a4-90cd-a61c408b040b' and name = '2';

delete from user_attribute where user_id = 'c7332040-acfd-41a4-90cd-a61c408b040b' and name = '1';
```

This leads to a DB error:

```
[2024-09-12 12:00:31] [40P01] ERROR: deadlock detected
[2024-09-12 12:00:31] Detail: Process 234 waits for ShareLock on transaction 21200; blocked by process 232.
[2024-09-12 12:00:31] Process 232 waits for ShareLock on transaction 21201; blocked by process 234.
[2024-09-12 12:00:31] Hint: See server log for query details.
[2024-09-12 12:00:31] Where: while deleting tuple (0,5) in relation "user_attribute"
```

### PR DESCRIPTION
**Pull request by ahus1**
This should reduce deadlocks on the user property table if the users are updated concurrently.

Closes #32852

### CODE REVIEW

**Comment by ahus1** [comment]
@sguilhen - this should resolve the deadlocks we've seen when the same user is re-imported concurrently in multiple threads. Please have a look when you have the time. Thanks!

File: federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/LDAPMappersComparator.java [hunk 1]
```diff
@@ -98,13 +93,21 @@ public int compare(ComponentModel o1, ComponentModel o2) {
...
-            return 0;
+            return compareWithStableOrdering(o1, o2);
+        }
+
+        /**
+         * Ensure a stable ordering, so the mappers are always executed in the same order.
+         * This can avoid database deadlocks as the mappers will modify attributes always in the same order.
+         */
+        private static int compareWithStableOrdering(ComponentModel o1, ComponentModel o2) {
+            return o1.getId().compareTo(o2.getId());
```
Comment (sguilhen) [inline federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/LDAPMappersComparator.java:110]: We might revisit this when working in improving the LDAP group mapper as we might want to be able to set a specific order when there's more than one mapper (e.g. 2 mappers mapping different group DNs into different group paths in Keycloak). But it is a topic for a separate discussion in the future.

**Review by sguilhen — APPROVED** [review]

File: federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/LDAPMappersComparator.java (hunk 1 above)
Comment (ahus1) [reply federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/LDAPMappersComparator.java:110]: Let's consider this a fallback to ensure stable sorting. If we have criteria that are of a higher priority, we can always use those before using this fallback.

**Review by mhajas — APPROVED** [review]
Approving based on the review from @sguilhen

**Pull request merged**